File: BiteRunr/shared/src/commonMain/kotlin/org/biterunr_studios/biterunr/ReusableFunctions/MapValidationErrors.kt
package org.biterunr_studios.biterunr.ReusableFunctions

import org.biterunr_studios.biterunr.Models.ErrorResponse
import org.biterunr_studios.biterunr.Models.PathElement

fun mapValidationErrors(
    errorResponse: ErrorResponse,
    handlers: Map<String, (String) -> Unit>
) {
    for (issue in errorResponse.error.issues) {
        val pathKey = issue.path.joinToString(".") { pathElement ->
            when (pathElement) {
                is PathElement.StringElement -> pathElement.value
                is PathElement.NumberElement -> "[${pathElement.value}]"
            }
        }

        val exactHandler = handlers[pathKey]
        if (exactHandler != null) {
            exactHandler(issue.message)
            continue
        }

        val simplifiedPath = pathKey.replace(Regex("""\[\d+]"""), "")
        val simplifiedHandler = handlers[simplifiedPath]
        if (simplifiedHandler != null) {
            simplifiedHandler(issue.message)
        }
    }
}

File: BiteRunr/iosApp/iosApp/ReusableFunctions/MapValidationErrors.swift
import Shared

func mapValidationErrors<T>(
    _ fetchResponse: FetchResponse<T>,
    handlers: [String: (String) -> Void]
) {
    guard let fetchError = fetchResponse.error else {
        return
    }
    
    for issue in fetchError.issues {
        let pathKey = issue.path.map { pathElement in
            if let stringElement = pathElement as? PathElement.StringElement {
                return stringElement.value
            } else if let numberElement = pathElement as? PathElement.NumberElement {
                return "[\(numberElement.value)]"
            } else {
                return "unknown"
            }

        }.joined(separator: ".")
        
        // Try exact match
        if let handler = handlers[pathKey] {
            handler(issue.message)
        } else {
            // Try partial match (e.g., "order_users.[0].user_id" → "order_users.user_id")
            let simplifiedPath = pathKey.replacingOccurrences(of: #"\[\d+\]"#, with: "", options: .regularExpression)
            if let handler = handlers[simplifiedPath] {
                handler(issue.message)
            }
        }
    }
}

